File: 01-content-factory/python/agents/topic_agent.py
"""
Topic & Angle Agent
Generates high-impact content topics based on brand voice and performance data
"""
from typing import List
from ..models import BrandVoice, ContentTopic, ContentType
from ..providers.claude import ClaudeProvider
from ..utils import safe_extract_json


class TopicAgent:
    """
    World-class growth marketer agent that generates content topics
    using brand voice database and performance insights.
    """
    
    SYSTEM_PROMPT = """You are a world-class growth marketer for {industry}.
Using the brand voice database and last 90 days performance, generate content ideas that will 3x engagement.

You understand:
- Content marketing best practices and viral content patterns
- SEO keyword research principles and search intent
- Social media engagement patterns across platforms
- B2B and B2C marketing strategies
- Industry-specific content angles that drive conversions

Always return valid JSON arrays with no additional text."""

    USER_PROMPT = """Generate {count} high-impact content topics for the following client:

**Client:** {client_name}
**Industry:** {industry}
**Brand Voice:** {tone}
**Target Audience:** {target_audience}
**Keywords to incorporate:** {keywords}
**Content Goals:** {content_goals}
**Past Winners:** {past_winners}

For each topic, provide:
1. A compelling, clickable title
2. A unique angle or hook that differentiates from competitors
3. Relevant keywords (3-5) for SEO
4. Which content types it's suitable for: {content_types}
5. Priority level (high/medium/low) based on likely engagement

Return ONLY a valid JSON array with this structure:
[
  {{
    "id": "topic_1",
    "title": "...",
    "angle": "...",
    "keywords": ["...", "..."],
    "content_types": ["blog", "linkedin"],
    "priority": "high"
  }}
]"""

    def __init__(self):
        self.llm = ClaudeProvider()
    
    async def generate_topics(
        self,
        brand_voice: BrandVoice,
        count: int = 12,
        content_types: List[ContentType] = None,
    ) -> List[ContentTopic]:
        """Generate content topics for the week"""
        
        if content_types is None:
            content_types = [ContentType.BLOG, ContentType.LINKEDIN, ContentType.TWITTER]
        
        system = self.SYSTEM_PROMPT.format(industry=brand_voice.industry)
        
        user = self.USER_PROMPT.format(
            count=count,
            client_name=brand_voice.client_name,
            industry=brand_voice.industry,
            tone=brand_voice.tone,
            target_audience=brand_voice.target_audience,
            keywords=", ".join(brand_voice.keywords),
            content_goals=", ".join(brand_voice.content_goals),
            past_winners=", ".join(brand_voice.past_winners) if brand_voice.past_winners else "N/A",
            content_types=", ".join([ct.value for ct in content_types]),
        )
        
        response = await self.llm.generate_json(system, user)
        topics_data = safe_extract_json(response, expect_array=True)
        
        topics = []
        for i, topic in enumerate(topics_data):
            topics.append(ContentTopic(
                id=topic.get("id", f"topic_{i+1}"),
                title=topic["title"],
                angle=topic["angle"],
                keywords=topic["keywords"],
                content_types=[ContentType(ct) for ct in topic["content_types"]],
                priority=topic.get("priority", "medium"),
            ))
        
        return topics

File: 01-content-factory/agents/topic-agent.ts
import { generateWithClaude } from "../integrations/anthropic";
import type { ClientBrief, ContentTopic, ContentType, AgentResponse } from "../types";

const SYSTEM_PROMPT = `You are an expert content strategist for digital agencies. Your role is to generate highly engaging, SEO-optimized content topics that will resonate with target audiences and drive business results.

You understand:
- Content marketing best practices
- SEO keyword research principles
- Social media engagement patterns
- B2B and B2C marketing strategies
- Industry-specific content angles

Always output valid JSON arrays.`;

export async function generateTopics(
  brief: ClientBrief,
  count: number = 5,
  contentTypes: ContentType[] = ['blog', 'linkedin', 'twitter', 'video_script']
): Promise<AgentResponse<ContentTopic[]>> {
  try {
    const userPrompt = `Generate ${count} high-impact content topics for the following client:

**Client:** ${brief.clientName}
**Industry:** ${brief.industry}
**Brand Voice:** ${brief.brandVoice}
**Target Audience:** ${brief.targetAudience}
**Keywords to incorporate:** ${brief.keywords.join(', ')}
**Content Goals:** ${brief.contentGoals.join(', ')}

For each topic, provide:
1. A compelling title
2. A unique angle or hook
3. Relevant keywords (3-5)
4. Which content types it's suitable for from this list: ${contentTypes.join(', ')}
   - IMPORTANT: For at least 2-3 topics, include "video_script" in contentTypes if it's in the allowed list
   - Video scripts work best for topics with strong visual potential or emotional narratives
5. Priority level (high/medium/low)

Return ONLY a valid JSON array with this structure:
[
  {
    "id": "topic_1",
    "title": "...",
    "angle": "...",
    "keywords": ["...", "..."],
    "contentTypes": ["blog", "linkedin", "video_script"],
    "priority": "high"
  }
]`;

    const response = await generateWithClaude(SYSTEM_PROMPT, userPrompt, {
      maxTokens: 2048,
      temperature: 0.8,
    });

    const jsonMatch = response.match(/\[[\s\S]*\]/);
    if (!jsonMatch) {
      throw new Error("Failed to parse topics JSON from response");
    }

    const topics: ContentTopic[] = JSON.parse(jsonMatch[0]);
    
    return {
      success: true,
      data: topics,
    };
  } catch (error: any) {
    return {
      success: false,
      error: error.message || "Failed to generate topics",
    };
  }
}
